graph/simple/weighted_undirected.go

```diff
@@ -0,0 +1,260 @@
+}
+
+// NewWeightedUndirectedGraph returns an WeightedUndirectedGraph with the specified self and absent
+// edge weight values.
+func NewWeightedUndirectedGraph(self, absent float64) *WeightedUndirectedGraph {
+	return &WeightedUndirectedGraph{
+		nodes: make(map[int64]graph.Node),
+		edges: make(map[int64]map[int64]graph.WeightedEdge),
+
+		self:   self,
+		absent: absent,
+
+		nodeIDs: newIDSet(),
+	}
+}
+
+// NewNode returns a new unique Node to be added to g. The Node's ID does
+// not become valid in g until the Node is added to g.
+func (g *WeightedUndirectedGraph) NewNode() graph.Node {
+	if len(g.nodes) == 0 {
+		return Node(0)
+	}
+	if int64(len(g.nodes)) == maxInt {
+		panic("simple: cannot allocate node: no slot")
+	}
+	return Node(g.nodeIDs.newID())
+}
+
+// AddNode adds n to the graph. It panics if the added node ID matches an existing node ID.
+func (g *WeightedUndirectedGraph) AddNode(n graph.Node) {
+	if _, exists := g.nodes[n.ID()]; exists {
+		panic(fmt.Sprintf("simple: node ID collision: %d", n.ID()))
+	}
+	g.nodes[n.ID()] = n
+	g.edges[n.ID()] = make(map[int64]graph.WeightedEdge)
+	g.nodeIDs.use(n.ID())
+}
+
+// RemoveNode removes n from the graph, as well as any edges attached to it. If the node
+// is not in the graph it is a no-op.
+func (g *WeightedUndirectedGraph) RemoveNode(n graph.Node) {
+	if _, ok := g.nodes[n.ID()]; !ok {
+		return
+	}
+	delete(g.nodes, n.ID())
+
+	for from := range g.edges[n.ID()] {
+		delete(g.edges[from], n.ID())
+	}
+	delete(g.edges, n.ID())
+
+	g.nodeIDs.release(n.ID())
+}
+
+// NewWeightedEdge returns a new Edge from the source to the destination node.
+func (g *WeightedUndirectedGraph) NewWeightedEdge(from, to graph.Node, weight float64) graph.WeightedEdge {
+	return &WeightedEdge{F: from, T: to, W: weight}
+}
+
+// SetWeightedEdge adds an edge from one node to another. If the nodes do not exist, they are added.
+// It will panic if the IDs of the e.From and e.To are equal.
+func (g *WeightedUndirectedGraph) SetWeightedEdge(e graph.WeightedEdge) {
+	var (
+		from = e.From()
+		fid  = from.ID()
+		to   = e.To()
+		tid  = to.ID()
+	)
+
+	if fid == tid {
+		panic("simple: adding self edge")
+	}
+
+	if !g.Has(from) {
+		g.AddNode(from)
+	}
+	if !g.Has(to) {
+		g.AddNode(to)
+	}
+
+	g.edges[fid][tid] = e
+	g.edges[tid][fid] = e
+}
+
+// RemoveEdge removes e from the graph, leaving the terminal nodes. If the edge does not exist
+// it is a no-op.
+func (g *WeightedUndirectedGraph) RemoveEdge(e graph.Edge) {
+	from, to := e.From(), e.To()
+	if _, ok := g.nodes[from.ID()]; !ok {
+		return
+	}
+	if _, ok := g.nodes[to.ID()]; !ok {
+		return
+	}
+
+	delete(g.edges[from.ID()], to.ID())
+	delete(g.edges[to.ID()], from.ID())
+}
+
+// Node returns the node in the graph with the given ID.
+func (g *WeightedUndirectedGraph) Node(id int64) graph.Node {
+	return g.nodes[id]
+}
+
+// Has returns whether the node exists within the graph.
+func (g *WeightedUndirectedGraph) Has(n graph.Node) bool {
```

Comment:
Also an unrelated issue to this PR, but I do believe we should look into changing the signature of `Has` to `Has(id int64) bool`. And that too, would be a separate PR.